wire-java-generator/src/main/java/com/squareup/wire/java/JavaGenerator.java

```diff
@@ -154,24 +155,47 @@
   });
 
   private final Schema schema;
+  private final ImmutableMap<ProtoType, TypeName> nameToProtoFieldsJavaName;
   private final ImmutableMap<ProtoType, TypeName> nameToJavaName;
+  private final ImmutableMap<ProtoType, Map.Entry<ClassName, String>> protoTypeToAdapter;
   private final boolean emitAndroid;
   private final boolean emitCompact;
 
-  private JavaGenerator(Schema schema, ImmutableMap<ProtoType, TypeName> nameToJavaName,
+  private JavaGenerator(Schema schema, ImmutableMap<ProtoType, TypeName> nameToProtoFieldsJavaName,
+      ImmutableMap<ProtoType, TypeName> nameToJavaName,
+      ImmutableMap<ProtoType, Map.Entry<ClassName, String>> protoTypeToAdapter,
       boolean emitAndroid, boolean emitCompact) {
     this.schema = schema;
+    this.nameToProtoFieldsJavaName = nameToProtoFieldsJavaName;
     this.nameToJavaName = nameToJavaName;
+    this.protoTypeToAdapter = protoTypeToAdapter;
     this.emitAndroid = emitAndroid;
     this.emitCompact = emitCompact;
   }
 
   public JavaGenerator withAndroid(boolean emitAndroid) {
-    return new JavaGenerator(schema, nameToJavaName, emitAndroid, emitCompact);
+    return new JavaGenerator(schema, nameToProtoFieldsJavaName, nameToJavaName, protoTypeToAdapter,
+        emitAndroid, emitCompact);
   }
 
   public JavaGenerator withCompact(boolean compactGeneration) {
-    return new JavaGenerator(schema, nameToJavaName, emitAndroid, compactGeneration);
+    return new JavaGenerator(schema, nameToProtoFieldsJavaName, nameToJavaName, protoTypeToAdapter,
+        emitAndroid, compactGeneration);
+  }
+
+  public JavaGenerator withCustomProtoAdapter(ImmutableMap<ProtoType, TypeName> nameToJavaName,
+      ImmutableMap<ProtoType, Map.Entry<ClassName, String>> protoTypeToAdapter) {
+    Map<ProtoType, TypeName> clone = new LinkedHashMap<>(this.nameToJavaName);
```

Comment:
nit: the word `clone` is a bit distracting here. There’s a method called `Object.clone()` that does a specific thing, and this isn’t it. Maybe you instead want to just shadow the target? Ie. name this `nameToJavaName` ?